### PR TITLE
[#27] Implement DB serialization

### PR DESCRIPTION
**Pull request by Sheptytskyid**

### CODE REVIEW

File: src/main/java/com/manardenza/dao/HotelDaoImpl.java [hunk 1]
```diff
@@ -2,13 +2,18 @@
...
 
+import java.io.File;
 import java.util.ArrayList;
 import java.util.List;
 
 public class HotelDaoImpl extends AbstractDao<Hotel> {
 
     private static HotelDaoImpl instance;
 
+    {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/HotelDaoImpl.java]: MAJOR Move the contents of this initializer to a standard constructor or to field initializers. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1171)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 2]
```diff
@@ -2,27 +2,63 @@
...
+            e.printStackTrace();
+        }
+
+    }
+
+    public void readFromFile() {
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(databaseFile))) {
+            database = (List<T>) ois.readObject();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 3]
```diff
@@ -2,27 +2,63 @@
...
+        } catch (ClassNotFoundException e) {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)
MINOR Combine this catch with the one at line 31, which has the same body. (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2147)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 4]
```diff
@@ -2,27 +2,63 @@
...
+            e.printStackTrace();
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 5]
```diff
@@ -2,27 +2,63 @@
...
 import java.util.List;
 
 public abstract class AbstractDao<T extends AbstractObject> {
 
     protected List<T> database = new ArrayList<>();
+    protected File databaseFile;
+
+    public void saveToFile() {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(databaseFile))) {
+            oos.writeObject(database);
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Make the "java.util.List" class "Serializable" or don't write it. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2118)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 6]
```diff
@@ -2,27 +2,63 @@
...
+        } catch (IOException e) {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 7]
```diff
@@ -2,27 +2,63 @@
...
+            e.printStackTrace();
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 8]
```diff
@@ -2,27 +2,63 @@
...
+        }
+
+    }
+
+    public void readFromFile() {
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(databaseFile))) {
+            database = (List<T>) ois.readObject();
+        } catch (IOException e) {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/UserDaoImpl.java [hunk 9]
```diff
@@ -2,12 +2,17 @@
...
 import com.manardenza.entity.User;
 
+import java.io.File;
 import java.util.List;
 
 public class UserDaoImpl extends AbstractDao<User> {
 
     private static UserDaoImpl instance;
 
+    {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/UserDaoImpl.java]: MAJOR Move the contents of this initializer to a standard constructor or to field initializers. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1171)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 10]
```diff
@@ -2,27 +2,63 @@
...
+import java.io.ObjectOutputStream;
 import java.util.ArrayList;
 import java.util.List;
 
 public abstract class AbstractDao<T extends AbstractObject> {
 
     protected List<T> database = new ArrayList<>();
+    protected File databaseFile;
+
+    public void saveToFile() {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: Make this private

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 11]
```diff
@@ -2,27 +2,63 @@
...
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(databaseFile))) {
+            oos.writeObject(database);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+    }
+
+    public void readFromFile() {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: Make this private

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 12]
```diff
@@ -2,27 +2,61 @@
...
+            e.printStackTrace();
+        }
+
+    }
+
+    private void readFromFile() {
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(databaseFile))) {
+            database = (List<T>) ois.readObject();
+        } catch (IOException | ClassNotFoundException e) {
+            e.printStackTrace();
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 13]
```diff
@@ -2,27 +2,61 @@
...
 import java.util.List;
 
 public abstract class AbstractDao<T extends AbstractObject> {
 
     protected List<T> database = new ArrayList<>();
+    protected File databaseFile;
+
+    private void saveToFile() {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(databaseFile))) {
+            oos.writeObject(database);
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Make the "java.util.List" class "Serializable" or don't write it. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2118)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 14]
```diff
@@ -2,27 +2,61 @@
...
+        } catch (IOException e) {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 15]
```diff
@@ -2,27 +2,61 @@
...
+            e.printStackTrace();
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 16]
```diff
@@ -2,27 +2,61 @@
...
+        }
+
+    }
+
+    private void readFromFile() {
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(databaseFile))) {
+            database = (List<T>) ois.readObject();
+        } catch (IOException | ClassNotFoundException e) {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/manardenza/dao/AbstractDao.java (hunk 12 above)
Comment (zozich) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: You will need to do this - log this exception with the logger. As logging branch is not merged yet, use //FIXME: comments not to forget to fix this issue in future

File: src/main/java/com/manardenza/dao/AbstractDao.java (hunk 14 above)
Comment (zozich) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: //FIXME:

File: src/main/java/com/manardenza/dao/AbstractDao.java (hunk 15 above)
Comment (zozich) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: //FIXME:

File: src/main/java/com/manardenza/dao/AbstractDao.java (hunk 16 above)
Comment (zozich) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: //FIXME:

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 17]
```diff
@@ -2,27 +2,61 @@
...
 import java.util.ArrayList;
 import java.util.List;
 
 public abstract class AbstractDao<T extends AbstractObject> {
 
     protected List<T> database = new ArrayList<>();
+    protected File databaseFile;
+
+    private void saveToFile() {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(databaseFile))) {
```
Comment (zozich) [inline src/main/java/com/manardenza/dao/AbstractDao.java:24]: Try storing ObjectOutputStream variable as an instance variable not to open new stream each time any database call is performed

**Review by zozich — CHANGES_REQUESTED** [review]
Please amend

File: src/main/java/com/manardenza/entity/Hotel.java [hunk 18]
```diff
@@ -2,10 +2,11 @@
 
 import lombok.Data;
 
+import java.io.Serializable;
 import java.util.List;
 
 @Data
-public class Hotel extends AbstractObject {
+public class Hotel extends AbstractObject implements Serializable {
```
Comment (zozich) [inline src/main/java/com/manardenza/entity/Hotel.java]: If you have marked AbstractObject as Serializable, this class will become so as well

File: src/main/java/com/manardenza/entity/Reservation.java [hunk 19]
```diff
@@ -3,11 +3,12 @@
 import lombok.AllArgsConstructor;
 import lombok.Data;
 
+import java.io.Serializable;
 import java.util.Date;
 
 @Data
 @AllArgsConstructor
-public class Reservation {
+public class Reservation implements Serializable {
```
Comment (zozich) [inline src/main/java/com/manardenza/entity/Reservation.java:11]: If you have marked AbstractObject as Serializable, this class will become so as well

File: src/main/java/com/manardenza/entity/Room.java [hunk 20]
```diff
@@ -2,8 +2,10 @@
 
 import lombok.Data;
 
+import java.io.Serializable;
+
 @Data
-public class Room extends AbstractObject {
+public class Room extends AbstractObject implements Serializable{
```
Comment (zozich) [inline src/main/java/com/manardenza/entity/Room.java]: If you have marked AbstractObject as Serializable, this class will become so as well

File: src/main/java/com/manardenza/entity/User.java [hunk 21]
```diff
@@ -2,8 +2,10 @@
 
 import lombok.Data;
 
+import java.io.Serializable;
+
 @Data
-public class User extends AbstractObject {
+public class User extends AbstractObject implements Serializable {
```
Comment (zozich) [inline src/main/java/com/manardenza/entity/User.java]: If you have marked AbstractObject as Serializable, this class will become so as well

**Review by zozich — CHANGES_REQUESTED** [review]
Amend

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 22]
```diff
@@ -2,27 +2,78 @@
...
+        } catch (IOException ex) { //FIXME: Either log or rethrow this exception.
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
+        }
+    }
+
+    private void readFromFile() {
+        try {
+            database = (List<T>) ois.readObject();
+            ois.close();
+        } catch (IOException | ClassNotFoundException ex) { //FIXME: Either log or rethrow this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 23]
```diff
@@ -2,27 +2,78 @@
...
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 24]
```diff
@@ -2,27 +2,78 @@
...
+            ois = new ObjectInputStream(fis);
+
+        } catch (IOException ex) { //FIXME: Either log or rethrow this exception.
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
+        }
+    }
+
+    private void saveToFile() {
+        try {
+            oos.writeObject(database);
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Make the "java.util.List" class "Serializable" or don't write it. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2118)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 25]
```diff
@@ -2,27 +2,78 @@
...
 public abstract class AbstractDao<T extends AbstractObject> {
 
     protected List<T> database = new ArrayList<>();
+    protected File databaseFile;
+    protected FileOutputStream fos = null;
+    protected ObjectOutputStream oos = null;
+    protected FileInputStream fis = null;
+    protected ObjectInputStream ois = null;
+
+    {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Move the contents of this initializer to a standard constructor or to field initializers. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1171)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 26]
```diff
@@ -2,27 +2,78 @@
...
+        } catch (IOException ex) { //FIXME: Either log or rethrow this exception.
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
+        }
+    }
+
+    private void saveToFile() {
+        try {
+            oos.writeObject(database);
+            oos.close();
+        } catch (IOException ex) { //FIXME: Either log or rethrow this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 27]
```diff
@@ -2,27 +2,78 @@
...
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 28]
```diff
@@ -2,27 +2,78 @@
...
+    protected ObjectInputStream ois = null;
+
+    {
+        try {
+            fos = new FileOutputStream(databaseFile);
+            oos = new ObjectOutputStream(fos);
+            fis = new FileInputStream(databaseFile);
+            ois = new ObjectInputStream(fis);
+
+        } catch (IOException ex) { //FIXME: Either log or rethrow this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 29]
```diff
@@ -2,27 +2,78 @@
...
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 14 issues
* ![MAJOR][MAJOR] 11 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/manardenza/dao/HotelDaoImpl.java [hunk 30]
```diff
@@ -1,14 +1,20 @@
 package com.manardenza.dao;
 
 import com.manardenza.entity.Hotel;
+import com.manardenza.utils.DatabaseSeeder;
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/HotelDaoImpl.java]: MINOR Remove this unused import 'com.manardenza.utils.DatabaseSeeder'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/manardenza/dao/HotelDaoImpl.java [hunk 31]
```diff
@@ -1,14 +1,20 @@
...
+import java.io.File;
+import java.util.ArrayList;
 import java.util.List;
-import java.util.stream.Collectors;
 
 public class HotelDaoImpl extends AbstractDao<Hotel> {
 
     private static HotelDaoImpl instance;
 
+    {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/HotelDaoImpl.java]: MAJOR Move the contents of this initializer to a standard constructor or to field initializers. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1171)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 32]
```diff
@@ -2,27 +2,78 @@
...
+            oos.writeObject(database);
+        } catch (IOException ex) { //FIXME: Either log or rethrow this exception.
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
+        }
+    }
+
+    private void readFromFile() {
+        try {
+            database = (List<T>) ois.readObject();
+        } catch (IOException | ClassNotFoundException ex) { //FIXME: Either log or rethrow this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 33]
```diff
@@ -2,27 +2,78 @@
...
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 34]
```diff
@@ -2,27 +2,78 @@
...
+
+        } catch (IOException ex) { //FIXME: Either log or rethrow this exception.
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
+        }
+    }
+
+    private void saveToFile() {
+        try {
+            oos.writeObject(database);
+        } catch (IOException ex) { //FIXME: Either log or rethrow this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 35]
```diff
@@ -2,27 +2,78 @@
...
+            ex.printStackTrace(); //FIXME: Use a logger to log this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/UserDaoImpl.java [hunk 36]
```diff
@@ -2,10 +2,17 @@
 
 import com.manardenza.entity.User;
 
+import java.io.File;
+import java.util.List;
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/UserDaoImpl.java]: MINOR Remove this unused import 'java.util.List'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/manardenza/dao/UserDaoImpl.java [hunk 37]
```diff
@@ -2,10 +2,17 @@
...
+
 public class UserDaoImpl extends AbstractDao<User> {
 
     private static UserDaoImpl instance;
 
+    {
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/UserDaoImpl.java]: MAJOR Move the contents of this initializer to a standard constructor or to field initializers. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1171)

File: src/main/java/com/manardenza/Main.java [hunk 38]
```diff
@@ -1,9 +1,10 @@
 package com.manardenza;
 
+import com.manardenza.utils.DatabaseSeeder;
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/Main.java]: MINOR Remove this unused import 'com.manardenza.utils.DatabaseSeeder'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 18 issues
* ![MAJOR][MAJOR] 12 major
* ![MINOR][MINOR] 6 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 39]
```diff
@@ -2,27 +2,60 @@
...
 import java.util.List;
 
 public abstract class AbstractDao<T extends AbstractObject> {
 
     protected List<T> database = new ArrayList<>();
+    protected File databaseFile;
+
+    private void saveToFile() {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(databaseFile))) {
+            oos.writeObject(database);
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java]: MAJOR Make the "java.util.List" class "Serializable" or don't write it. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2118)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 40]
```diff
@@ -2,27 +2,60 @@
...
+        } catch (IOException e) { //FIXME: Either log or rethrow this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java:26]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 41]
```diff
@@ -2,27 +2,60 @@
...
+            e.printStackTrace();//FIXME: Use a logger to log this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java:27]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 42]
```diff
@@ -2,27 +2,60 @@
...
+        }
+    }
+
+    private void readFromFile() {
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(databaseFile))) {
+            database = (List<T>) ois.readObject();
+        } catch (IOException | ClassNotFoundException e) { //FIXME: Either log or rethrow this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java:34]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 43]
```diff
@@ -2,27 +2,60 @@
...
+            e.printStackTrace();//FIXME: Use a logger to log this exception.
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java:35]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1134)
MINOR Use a logger to log this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1148)

**Comment by Sheptytskyid** [comment]
SonarQube analysis reported 9 issues
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Review by zozich — APPROVED** [review]
Cool!

File: src/main/java/com/manardenza/dao/AbstractDao.java [hunk 44]
```diff
@@ -2,27 +2,64 @@
...
     protected List<T> database = new ArrayList<>();
+    private File databaseFile;
+
+    protected AbstractDao(File databaseFile) {
+        this.databaseFile = databaseFile;
+    }
+
+    private void saveToFile() {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(databaseFile))) {
+            oos.writeObject(database);
```
Comment (Sheptytskyid) [inline src/main/java/com/manardenza/dao/AbstractDao.java:25]: MAJOR Make the "java.util.List" class "Serializable" or don't write it. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2118)

**Pull request merged**